Add certificate support for Azure Batch (#3097)
terraform-providers/terraform-provider-azurerm: website/docs/r/batch_certificate.html.markdown

---
layout: "azurerm"
page_title: "Azure Resource Manager: azurerm_batch_certificate"
sidebar_current: "docs-azurerm-resource-batch-certificate"
description: |-
  Manages a certificate in an Azure Batch account.

---

> # azurerm_batch_account

Reviewer: ```suggestion
# azurerm_batch_certificate
```

  resource_group_name  = "${azurerm_resource_group.test.name}"
  location             = "${azurerm_resource_group.test.location}"
  pool_allocation_mode = "BatchService"
  storage_account_id   = "${azurerm_storage_account.test.id}"

  tags = {
    env = "test"
  }
}

resource "azurerm_batch_certificate" "test" {
  resource_group_name  = "${azurerm_resource_group.test.name}"
  account_name         = "${azurerm_batch_account.test.name}"
  certificate          = "${base64encode(file("certificate.pfx"))}"
  format               = "Pfx"
  password             = "terraform"
  thumbprint           = "42C107874FD0E4A9583292A2F1098E8FE4B2EDDA"
  thumbprint_algorithm = "SHA1"
}
```

## Argument Reference

The following arguments are supported:

* `account_name` - (Required) Specifies the name of the Batch account. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) The name of the resource group in which to create the Batch account. Changing this forces a new resource to be created.

* `certificate` - (Required) The base64-encoded contents of the certificate.

* `format` - (Required) The format of the certificate. Possible values are `Cer` or `Pfx`.

> * `password` - (Optional) The password to access the certificate's private key. This must and can only be specified when `format` is `Pfx`.

Reviewer: above the comment says:
>  Required if `format` is "Cer"

I'm assuming one of these is valid but not the other - although if they are should we make this field Required?

Author: I'm not sure which bit had the "Required if format is 'Cer'" comment - can you point me to it?

Reviewer: no worries - sure: https://github.com/terraform-providers/terraform-provider-azurerm/pull/3097/files#diff-6bcfbf3bca66879558f904216ebf7254R54

Author: Got it - fixed now!

> * `thumbprint` - (Required) The thumbprint of the certificate. This must match the thumbprint from the name.

Reviewer: Since we've removed the `name` field (and infer it) we can make this:

```suggestion
* `thumbprint` - (Required) The thumbprint of the certificate.
```

> * `thumbprint_algorithm` - (Required) The algorithm of the certificate thumbprint. This must match the first portion of the certificate name. Currently required to be 'SHA1'.

Reviewer: _minor_ for consistency with other resources we'd generally make this:

```suggestion
* `thumbprint_algorithm` - (Required) The algorithm of the certificate thumbprint. At this time the only supported value is `SHA1`.
```

  name                     = "teststorage"
  resource_group_name      = "${azurerm_resource_group.test.name}"
  location                 = "${azurerm_resource_group.test.location}"
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_batch_account" "test" {
  name                 = "testbatchaccount"
  resource_group_name  = "${azurerm_resource_group.test.name}"
  location             = "${azurerm_resource_group.test.location}"
  pool_allocation_mode = "BatchService"
  storage_account_id   = "${azurerm_storage_account.test.id}"

  tags = {
    env = "test"
  }
}

resource "azurerm_batch_certificate" "test" {
  name                 = "SHA1-42C107874FD0E4A9583292A2F1098E8FE4B2EDDA"
  resource_group_name  = "${azurerm_resource_group.test.name}"
  account_name         = "${azurerm_batch_account.test.name}"
  certificate          = "${base64encode(file("certificate.pfx"))}"
  format               = "Pfx"
  password             = "terraform"
  thumbprint           = "42C107874FD0E4A9583292A2F1098E8FE4B2EDDA"
  thumbprint_algorithm = "SHA1"
}
```

## Argument Reference

The following arguments are supported:

> * `name` - (Required) The identifier for the certificate. This must be made up of algorithm and thumbprint separated by a dash, and must match the certificate data in the request. For example SHA1-a3d1c5.

Reviewer: we traditionally quote values/examples - also this is a ForceNew, so can we make this

```suggestion
* `name` - (Required) The identifier for the certificate. This must be made up of algorithm and thumbprint separated by a dash, and must match the certificate data in the request. For example `SHA1-a3d1c5`. Changing this forces a new resource to be created.
```

* `account_name` - (Required) Specifies the name of the Batch account. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) The name of the resource group in which to create the Batch account. Changing this forces a new resource to be created.

* `certificate` - (Required) The base64-encoded contents of the certificate.

> * `format` - (Optional) The format of the certificate - either Pfx or Cer. If omitted, the default is Pfx.

Reviewer: can we quote these values:

```suggestion
* `format` - (Required) The format of the certificate. Possible values are `Cer` or `Pfx`.
```

(I've assumed this is now Required here, as per the comment above)

> * `password` - The password to access the certificate's private key. This is required if the certificate format is pfx and must be omitted if the certificate format is cer.

Reviewer: we phrase this as:

```suggestion
* `password` - (Optional) The password to access the certificate's private key. This must and can only be specified when `format` is `Pfx`.
```

> * `thumbprint` - (Required) The thumbprint of the certificate. This must match the thumbprint from the name.

Reviewer: given the name is built up from these two values, I feel like it a better UX to remove the `name` field and infer it from these two values, WDYT?

Author: Done!
